sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala

```diff
@@ -344,6 +344,11 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
         table, provider, partitionColumnNames, bucketSpec, mode, options, query)
     } else {
       val struct = Option(ctx.colTypeList()).map(createStructType)
+      if (struct.isEmpty && bucketSpec.nonEmpty) {
```

Comment:
In that case, partition columns are inferred at runtime. Thus, although we ignore user specified partition columns, but we still treat the table as a partition table. 

Here, we will really ignore the bucket spec, because the bucket columns are not inferred at runtime. That means, we will not treat it as a bucket table. That is the major reason why this PR issues the exception.

Let me know if Log message is still better here. Thanks!